A fix for BUKKIT-227.

wilsonsd opened the pull request:
Modified the "give" command to allow specifying data values.  E.g.
give user 17:2
will give the user named "user" a block of birch (rather than a block
of plain wood).  See
https://bukkit.atlassian.net/browse/BUKKIT-227

wilsonsd committed:
A fix for BUKKIT-227.

Modified the "give" command to allow specifying data values.  E.g.
give user 17:2
will give the user named "user" a block of birch (rather than a block
of plain wood).  See
https://bukkit.atlassian.net/browse/BUKKIT-227

grum commented:
This doesn't compile on 1.5 because of the Arrays.copyOfRange which is 1.6

wilsonsd committed:
Use System.arraycopy instead of Arrays.copyOfRange.

For Java 1.5 compatibility.

wilsonsd commented:
Okay, here's a fix compatible with 1.5.  I didn't see a version of the jdk specified anywhere, so I just assumed 1.6 based on this:
http://wiki.bukkit.org/Setting_Up_Your_Workspace#Java

Maybe a note to use 1.5 could be added to the Readme?  Or maybe the pom.xml file could specify a version of the jdk to use?  See
http://maven.apache.org/plugins/maven-compiler-plugin/examples/compile-using-different-jdk.html

tips48 commented:
How about org.bukkit.util.Java15Compat.Arrays_CopyOf() ?

feildmaster commented:
@wilsonsd : Added a note. ^^

wilsonsd committed:
Correct /give command syntax, used Java15Compat.

Modified the regex in GiveCommand to accept the same syntax as the
vanilla server.

Also switched System.arraycopy to Java15Compat.Arrays_copyOfRange

EvilSeph commented:
This is currently under review. It looks like this gets rid of the support for providing material names to the /give command and I'm not sure we want that.

wilsonsd committed:
Allow material names in give command.

Previous commits in this branch required users of the give command to
specify the item to give using its number id.  This commit lets users
specify the item to give using its number id or its name.  For example,
give user 5
and
give user wood
will now both give the user a block of wood.

wilsonsd commented on src/main/java/org/bukkit/command/defaults/GiveCommand.java:
```diff
@@ -7,45 +7,84 @@
-                Command.broadcastCommandMessage(sender, "Giving " + player.getName() + " some " + material.getId() + " (" + material + ")");
-
+            String argString = join(Java15Compat.Arrays_copyOfRange(args, 1, args.length), " ");
+            Pattern pattern = Pattern.compile("(\\w+) *(:|[0-9]+)? *(:|[0-9]+)?");
```
Changed the regex from "([0-9]+) ".... to "(\w+)"....  This lets the item to give be specified by name or id number.

Dinnerbone commented:
Is this still needed? Don't we already have support?

Edit: Apparently it's not and we do. Closing off. Thanks anyhows!
